### ISSUE TITLE
XML Fault serialization fails if an xsi:type contains a schema identifier

### ISSUE THREAD

**Issue by atanasa** [dotnet/wcf#2541]
If the server replies with the following SOAP fault:

```xml
<soapenv:Envelope xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/"
		xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"
		xmlns:xsd="http://www.w3.org/2001/XMLSchema"
		xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<soapenv:Body>
		<soapenv:Fault>
			<faultcode>ServerFaultCode</faultcode>
			<faultstring>The resource &apos;Storage1&apos; is in use.</faultstring>
			<detail>
				<ResourceInUseFault xmlns="urn:vim25" xsi:type="ResourceInUse">
					<faultMessage>
						<key>...</key>
						<arg>
							<key>name</key>
							<value xsi:type="xsd:string">Storage1</value>
						</arg>
						<arg>
							<key>reason</key>
							<value xsi:type="xsd:string">vmsRegistered</value>
						</arg>
						<message>...</message>
					</faultMessage>
					<name>Storage1</name>
				</ResourceInUseFault>
			</detail>
		</soapenv:Fault>
	</soapenv:Body>
</soapenv:Envelope>
```
the client fault serialization would fail with the error: 
Namespace prefix 'xsd' is not defined.
```
   at System.Xml.Serialization.XmlSerializationReader.ToXmlQualifiedName(String value, Boolean decodeName)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationReaderResourceInUse.Read1_Object(Boolean isNullable, Boolean checkType)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationReaderResourceInUse.Read4_KeyAnyValue(Boolean isNullable, Boolean checkType)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationReaderResourceInUse.Read5_LocalizableMessage(Boolean isNullable, Boolean checkType)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationReaderResourceInUse.Read8_ResourceInUse(Boolean isNullable, Boolean checkType)
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationReaderResourceInUse.Read9_ResourceInUseFault()

   at System.Xml.Serialization.XmlSerializer.Deserialize(XmlReader xmlReader, String encodingStyle, XmlDeserializationEvents events)
   at System.ServiceModel.Dispatcher.XmlSerializerObjectSerializer.ReadObject(XmlDictionaryReader reader, Boolean verifyObjectName)
   at System.Runtime.Serialization.XmlObjectSerializer.ReadObjectHandleExceptions(XmlReaderDelegator reader, Boolean verifyObjectName, DataContractResolver dataContractResolver)
   at System.Runtime.Serialization.XmlObjectSerializer.ReadObject(XmlDictionaryReader reader)
   at System.ServiceModel.Dispatcher.XmlSerializerFaultFormatter.CreateFaultException(MessageFault messageFault, String action)
   at System.ServiceModel.Channels.ServiceChannel.HandleReply(ProxyOperationRuntime operation, ProxyRpc& rpc)
   at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs, TimeSpan timeout)
   at System.ServiceModel.Channels.ServiceChannelProxy.InvokeService(IMethodCallMessage methodCall, ProxyOperationRuntime operation)
   at System.ServiceModel.Channels.ServiceChannelProxy.Invoke(IMessage message)
```
The value element is defined in the xsd schema as:
```xml
<element name="value" type="xsd:anyType" />
```

Looking at the code generated for the deserializer, I am under the impression that the deserializer would try to read the xsi:type attribute to determine the actual type. In this case the value of the xsi:type is "xsd:string", where xsd is defined at the root element of the SOAP envelope.
I think this is the expected behavior.
I think that then the call to ToXmlQualifiedName on the "xsd:string" value fails because the reader does not have a definition for the xsd namespace.

Looking at the code of the MessageFault.GetReaderAtDetailContents, it looks like the fault details element is copied to a buffer and a reader for that buffer is returned. 

The issue is probably that this way the schema definitions of the original xml reader are lost if those namespaces are not present in the xml fragment that is being extracted.

To summarize,
The xml SOAP response is correct, but the WCF fault deserializer fails to deserialize the fault details. (Probably due to losing the schema definitions information.)

**Comment by zhenlan** [comment]
@atanasa thanks for reporting the issue. Do you know if this issue repro's on the full framework as well?

**Comment by atanasa** [comment]
Yes, it reproduces on the .net framework too.

**Comment by zhenlan** [comment]
@atanasa, thanks for the confirmation. 

@huanwu, can you please take a look?

**Comment by huanwu** [comment]
@atanasa Could you share with me your repro?

**Comment by atanasa** [comment]
Here is some code to reproduce this in isolation:
[ReproWcfErrorDeserialization.zip](https://github.com/dotnet/wcf/files/1825701/ReproWcfErrorDeserialization.zip)

Steps to reproduce:
1. Run the project DummyServer (this would host a HTTP server at 10888 that simulates the server)
2. Run the project WcfDummyClient that makes a simple call and tries to handle the error.

Expected:
The client should raise a FaultException of DummyFault detail.

Actual:
Got a generic exception that the XML document is not valid (Namespace prefix 'xsd' is not defined.)

Both projects are on .net core 2.0. Retargetting the client to net4* should be able to reproduce it for .Net Framework.

**Comment by huanwu** [comment]
@atanasa Thanks! I will take a look.

**Comment by huanwu** [comment]
@atanasa you need specify xmlns:xsd on DumyFault element directly as the following. Could you try?

<DummyFault xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
          <SomeProperty xsi:type="xsd:string">Dummy text</SomeProperty>
        </DummyFault>

**Comment by atanasa** [comment]
No, I don't control the server and can't change the responce. As far as I know, the responce is valid XML and SOAP - so WCF should be able to process it.

**Comment by huanwu** [comment]
@atanasa I will take a look.

**Comment by HongGit** [comment]
@mconnew  could you please take a look?

**Comment by atanasa** [comment]
Looking at the code - `MessageFault.CreateFault12Driver` and `MessageFault.CreateFault12Driver`, the issue seems to be that `XmlDictionaryWriter.WriteNode` is not preserving the namespaces when coping the fault into the buffer.
I have filed an issue with .Net for that: https://github.com/dotnet/runtime/issues/49010
